Zend/zend_fibers.c

```diff
@@ -0,0 +1,714 @@
+
+	zend_fiber_suspend_context(&fiber->context);
+
+	ZEND_FIBER_RESTORE_EG(stack, stack_page_size, execute_data, error_reporting, jit_trace_num);
+}
+
+static void zend_fiber_switch_to(zend_fiber *fiber)
+{
+	zend_fiber *previous;
+	zend_vm_stack stack;
+	size_t stack_page_size;
+	zend_execute_data *execute_data;
+	int error_reporting;
+	uint32_t jit_trace_num;
+
+	previous = EG(current_fiber);
+
+	zend_observer_fiber_switch_notify(previous, fiber);
+
+	ZEND_FIBER_BACKUP_EG(stack, stack_page_size, execute_data, error_reporting, jit_trace_num);
+
+	EG(current_fiber) = fiber;
+
+	zend_fiber_switch_context(&fiber->context);
+
+	EG(current_fiber) = previous;
+
+	ZEND_FIBER_RESTORE_EG(stack, stack_page_size, execute_data, error_reporting, jit_trace_num);
+
+	zend_observer_fiber_switch_notify(fiber, previous);
+
+	if (UNEXPECTED(EG(fiber_error)) && fiber->status != ZEND_FIBER_STATUS_SHUTDOWN) {
+		if (previous) {
+			zend_fiber_suspend(previous); // Still in fiber, suspend again until in {main}.
+			abort(); // This fiber should never be resumed.
+		}
+
+		zend_fiber_error *error = EG(fiber_error);
+		zend_error_zstr_at(error->type, error->filename, error->lineno, error->message);
+	}
+}
+
+ZEND_COLD void zend_error_suspend_fiber(
+		int orig_type, zend_string *error_filename, uint32_t error_lineno, zend_string *message)
+{
+	ZEND_ASSERT(EG(current_fiber) && "Must be within an active fiber!");
+
+	zend_fiber_error *error = emalloc(sizeof(zend_fiber_error));
+
+	error->type = orig_type;
+	error->filename = error_filename;
+	error->lineno = error_lineno;
+	error->message = message;
+
+	EG(fiber_error) = error;
+
+	zend_fiber_suspend(EG(current_fiber));
+
+	abort(); // This fiber should never be resumed.
+}
+
+static zend_always_inline zend_vm_stack zend_fiber_vm_stack_alloc(size_t size)
+{
+	zend_vm_stack page = emalloc(size);
+
+	page->top = ZEND_VM_STACK_ELEMENTS(page);
+	page->end = (zval *) ((uintptr_t) page + size);
+	page->prev = NULL;
+
+	return page;
+}
+
+static void ZEND_STACK_ALIGNED zend_fiber_execute(zend_fiber_context *context)
+{
+	zend_fiber *fiber = EG(current_fiber);
+	ZEND_ASSERT(fiber);
+
+	zend_long error_reporting = INI_INT("error_reporting");
+	if (!error_reporting && !INI_STR("error_reporting")) {
+		error_reporting = E_ALL;
+	}
+
+	zend_vm_stack stack = zend_fiber_vm_stack_alloc(ZEND_FIBER_VM_STACK_SIZE);
+	EG(vm_stack) = stack;
+	EG(vm_stack_top) = stack->top + ZEND_CALL_FRAME_SLOT;
+	EG(vm_stack_end) = stack->end;
+	EG(vm_stack_page_size) = ZEND_FIBER_VM_STACK_SIZE;
+
+	fiber->execute_data = (zend_execute_data *) stack->top;
+
+	memset(fiber->execute_data, 0, sizeof(zend_execute_data));
+
+	EG(current_execute_data) = fiber->execute_data;
+	EG(jit_trace_num) = 0;
+	EG(error_reporting) = error_reporting;
+
+	fiber->fci.retval = &fiber->value;
+
+	// Reference added while running, removed when suspended, and added again once resumed.
+	GC_ADDREF(&fiber->std);
```

Comment:
I tried dropping all these GC_ADDREF/GC_DELREFs and didn't get any test failures ... so either they aren't necessary, or we're missing test coverage. I'm not sure what one it is :/

Reply:
Seems the extra references aren't necessary with the current API – they definitely were necessary with the FiberScheduler API.

With the current API, I don't think there is a way to enter the fiber without having an external reference to the fiber, and I don't see a way for that reference to be dropped the fiber is running, so no need to add another reference.

Reply:
I also tried a reference from a global that was unset during fiber execution, but that was not an issue since the VM adds a reference to the object during the method call, so effective the VM is already doing what I was attempting to do here manually.